src/librustc_builtin_macros/source_util.rs

```diff
@@ -108,8 +108,7 @@ pub fn expand_include<'cx>(
             return DummyResult::any(sp);
         }
     };
-    let directory_ownership = DirectoryOwnership::Owned { relative: None };
-    let p = new_sub_parser_from_file(cx.parse_sess(), &file, directory_ownership, None, sp);
+    let p = new_sub_parser_from_file(cx.parse_sess(), &file, None, sp);
```

Comment:
This apparently broke Servo's `script` crate, or at least the copy on perf.rust-lang.org.
cc @Mark-Simulacrum can you post some details?

IIUC, this can't be fixed locally either, because the `mod bar;` parsed by `include!("some/other/dir/foo.rs");` will be expanded later, so for expanding `mod bar;` to load `some/other/dir/foo.rs` (which is apparently the old behavior? how are we not testing for it), we'd need to detect that an `include!` happened, or something similar.

Reply:
cc @petrochenkov

Reply:
This is a minimal reproduction:

```
$ tree .
.
|-- directory
|   |-- bar.rs
|   `-- included.rs
`-- foo.rs
```

```
// foo.rs
include!("directory/included.rs");

fn main() {}

// directory/included.rs
mod bar;

// directory/bar.rs
// empty
```

```
$ rustc +3c6f982cc908aacc39c3ac97f31c989f81cc213c foo.rs
error[E0583]: file not found for module `bar`
 --> directory/included.rs:1:1
  |
1 | mod bar;
  | ^^^^^^^^
  |
  = help: to create the module `bar`, create file "bar.rs"

error: aborting due to previous error

For more information about this error, try `rustc --explain E0583`.
```

Reply:
I've just recalled a related issue - https://github.com/rust-lang/rust/issues/58818.
Could you check whether this PR fixed it?

Reply:
>we'd need to detect that an include! happened, or something similar.

FWIW, we have `Definitions::expansions_that_defined` that can link from the `mod bar;` item to the macro that produced it. Then we have `ExpnData::parent` that can link from that macro to the macro that produced it (recursively).

Perhaps using these data we can hack up something reproducing the old behavior.

Reply:
We can only detect `include` comparing the macro's name in `ExpnData` with `sym::include`.
`fn expansion_cause` used in for a bunch of various stuff also hard-codes the name `include` like this, so it should be ok for now.

(The proper fix is a flag in `struct SyntaxExtension`, other file-including built-in macros should have it as well.)

Reply:
@petrochenkov So I tried two different approaches to set `ownership` before loading the external module. The first one was adding `Incovation::is_current_expansion_macro_named` defined like so:
```rust
    fn is_current_expansion_macro_named(&self, sym: Symbol) -> bool {
        let mut curr = self.cx.current_expansion.id;
        loop {
            let expn_data = curr.expn_data();
            if expn_data.kind == ExpnKind::Macro(MacroKind::Bang, sym) {
                // Stop going up the backtrace once `sym` is encountered.
                return true;
            }
            if expn_data.is_root() {
                // Reached root, need to bail to ensure termination.
                return false;
            }
            curr = expn_data.call_site.ctxt().outer_expn();
        }
    }
```
and then using it in `flat_map_item` before `parse_external_mod` happens:
```rust
                let pushed = &mut false; // Record `parse_external_mod` pushing so we can pop.
                let mut dir = Directory { ownership: orig_ownership, path: module.directory };

                if self.is_current_expansion_macro_named(sym::include) {
                    log::debug!("flat_map_item, is_current_expansion_macro_named = true");
                    dir.ownership = DirectoryOwnership::Owned { relative: None };
                }
```
The log message here does trigger on @Mark-Simulacrum's regression test, but this doesn't fix the issue.

I also tweaked `SyntaxExtension`, adding a flag `foobar: bool` which is used in `fully_expand_fragment`:
```rust
            let (expanded_fragment, new_invocations) = match res {
                InvocationRes::Single(ext) => match self.expand_invoc(invoc, &ext.kind) {
                    ExpandResult::Ready(fragment) => {
                        if ext.foobar {
                            self.cx.current_expansion.directory_ownership =
                                DirectoryOwnership::Owned { relative: None };
                            log::debug!("fully_expand_fragment, using foobar");
                        }
                        self.collect_invocations(fragment, &[])
```
and which is set in `compile_macro`:
```rust
        if result.is_builtin {
            // The macro was marked with `#[rustc_builtin_macro]`.
            if let Some(ext) = self.builtin_macros.remove(&item.ident.name) {
                // The macro is a built-in, replace its expander function
                // while still taking everything else from the source code.
                result.kind = ext.kind;

                if item.ident.name == sym::include {
                    result.foobar = true;
                    log::debug!("compile_macro, setting foobar");
                }
```
The log message also triggers here, but also does nothing to fix the issue, as the approach is the same as the other mechanism.

I'm not sure why it doesn't work. :(

Reply:
Found a fix, PR up at https://github.com/rust-lang/rust/pull/70184.